boot/bootutil/zephyr/src/boot_request_retention.c

```diff
@@ -0,0 +1,341 @@
+    size_t req_slot;
+    int ret;
+
+    ret = boot_request_slot_find(BOOT_REQUEST_IMG_PREFERENCE, image, &req_slot);
+    if (ret != 0) {
+        return ret;
+    }
+
+    switch (slot) {
+    case 0:
+        value = BOOT_REQUEST_SLOT_PRIMARY;
+        break;
+    case 1:
+        value = BOOT_REQUEST_SLOT_SECONDARY;
+        break;
+    default:
+        return -EINVAL;
+    }
+
+    return retention_write(bootloader_request_dev, req_slot * sizeof(value), (void *)&value,
+                           sizeof(value));
+}
+
+uint32_t
+boot_request_get_preferred_slot(uint8_t image)
+{
+    uint8_t value = BOOT_REQUEST_SLOT_INVALID;
+    size_t req_slot;
+    int ret;
+
+    ret = boot_request_slot_find(BOOT_REQUEST_IMG_PREFERENCE, image, &req_slot);
+    if (ret != 0) {
+        return BOOT_REQUEST_NO_PREFERRED_SLOT;
+    }
+
+    ret = retention_read(bootloader_request_dev, req_slot * sizeof(value), (void *)&value,
+                         sizeof(value));
+    if (ret != 0) {
+        return BOOT_REQUEST_NO_PREFERRED_SLOT;
+    }
+
+    switch (value) {
+    case BOOT_REQUEST_SLOT_PRIMARY:
+        return 0;
+    case BOOT_REQUEST_SLOT_SECONDARY:
+        return 1;
+    default:
+        break;
+    }
+
+    return BOOT_REQUEST_NO_PREFERRED_SLOT;
+}
+
+int
+boot_request_enter_recovery(void)
+{
+    uint8_t value = BOOT_REQUEST_MODE_RECOVERY;
+    size_t req_slot;
+    int ret;
+
+    ret = boot_request_slot_find(BOOT_REQUEST_BOOT_MODE, BOOT_REQUEST_IMG_BOOTLOADER,
+                                 &req_slot);
+    if (ret != 0) {
+        return ret;
+    }
+
+    return retention_write(bootloader_request_dev, req_slot * sizeof(value), (void *)&value,
+                           sizeof(value));
+}
+
+bool
+boot_request_detect_recovery(void)
+{
+    uint8_t value = BOOT_REQUEST_MODE_INVALID;
+    size_t req_slot;
+    int ret;
+
+    ret = boot_request_slot_find(BOOT_REQUEST_BOOT_MODE, BOOT_REQUEST_IMG_BOOTLOADER,
+                                 &req_slot);
+    if (ret != 0) {
+        return false;
+    }
+
+    ret = retention_read(bootloader_request_dev, req_slot * sizeof(value), (void *)&value,
+                         sizeof(value));
+    if ((ret == 0) && (value == BOOT_REQUEST_MODE_RECOVERY)) {
+        return true;
+    }
+
+    return false;
+}
+
+int
+boot_request_enter_firmware_loader(void)
+{
+    uint8_t value = BOOT_REQUEST_MODE_FIRMWARE_LOADER;
+    size_t req_slot;
+    int ret;
+
+    ret = boot_request_slot_find(BOOT_REQUEST_BOOT_MODE, BOOT_REQUEST_IMG_BOOTLOADER,
+                                 &req_slot);
+    if (ret != 0) {
+        return ret;
+    }
+
+    return retention_write(bootloader_request_dev, req_slot * sizeof(value), (void *)&value,
+                           sizeof(value));
+}
+
+bool
+boot_request_detect_firmware_loader(void)
+{
+    uint8_t value = BOOT_REQUEST_MODE_INVALID;
+    size_t req_slot;
+    int ret;
+
+    ret = boot_request_slot_find(BOOT_REQUEST_BOOT_MODE, BOOT_REQUEST_IMG_BOOTLOADER,
```

Comment:
#ifdef these with the firmware loader Kconfig or define

Reply:
Honestly, I do not see much value in adding those `ifdef`s - unreferenced functions will get lost and do not increase the build size and adding those create new inter-module dependencies, that may become painful later.

At this stage I will add them, because we do not have a second use case for the module, so it is hard to judge which option is better.